File: app/src/main/java/indi/yume/app/passwordbox/view/PwdCheckView.java
package indi.yume.app.passwordbox.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.annotation.IntRange;
import android.support.annotation.Size;
import android.support.annotation.StyleableRes;
import android.util.AttributeSet;
import android.view.View;

import indi.yume.app.passwordbox.R;
import indi.yume.app.passwordbox.util.CheckPwdUtil;

/**
 * Created by bush2 on 2016/4/5.
 */
public class PwdCheckView extends View {
    private static final int MAX_LEVEL = 5;
    private static final int[] RECT_COLOR_LIST = {
            0xffe11c1b,
            0xffe0791b,
            0xffe1ce1b,
            0xff83e01b,
            0xff1ce11a
    };
    private static final int BACK_RECT_COLOR = 0xffeeeeee;

    @Size(5)
    private Paint[] rectPaintList = {
            new Paint(),
            new Paint(),
            new Paint(),
            new Paint(),
            new Paint()
    };
    private Paint backRectPaint = new Paint();

    @Size(5)
    private Rect[] rectList = {
            new Rect(),
            new Rect(),
            new Rect(),
            new Rect(),
            new Rect()
    };
    private int level = 0;

    public PwdCheckView(Context context) {
        this(context, null);
    }

    public PwdCheckView(Context context, AttributeSet attrs) {
        super(context, attrs);

        if(attrs != null)
            init(context.obtainStyledAttributes(attrs, R.styleable.PwdCheckView));

        init();
    }

    private void init(TypedArray tArray) {
        int count = tArray.getIndexCount();
        for (int i = 0; i < count; i++)
            setAttr(tArray.getIndex(i), tArray);
        tArray.recycle();
    }

    private void setAttr(@StyleableRes int attrIndex, TypedArray tArray) {
        switch (attrIndex) {
            case R.styleable.PwdCheckView_pcv_password:
                String pwd = tArray.getString(attrIndex);
                checkLevel(pwd);
                break;
        }
    }

    private void init() {
        for (int i = 0; i < 5; i++) {
            rectPaintList[i].setAntiAlias(true);
            rectPaintList[i].setColor(RECT_COLOR_LIST[i]);
        }

        backRectPaint.setAntiAlias(true);
        backRectPaint.setColor(BACK_RECT_COLOR);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);

        int cellWithSpace = w / MAX_LEVEL;
        int space = cellWithSpace / 6;

        int cellWidth = (w - (MAX_LEVEL - 1) * space) / MAX_LEVEL;

        int xOffset = 0;
        for(int i = 0; i < MAX_LEVEL; i++){
            Rect rect = rectList[i];
            rect.set(xOffset, 0, xOffset + cellWidth, h);
            xOffset += cellWidth + space;
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        Paint paint;
        for(int i = 0; i < MAX_LEVEL; i++) {
            if(i > level)
                paint = backRectPaint;
            else
                paint = rectPaintList[i];

            canvas.drawRect(rectList[i], paint);
        }
    }

    public void setLevel(@IntRange(from = 0, to = MAX_LEVEL) int level) {
        this.level = level;
        postInvalidate();
    }

    public void checkLevel(String pwd) {
        this.level = CheckPwdUtil.checkPwdStrong(pwd, 8, 8);
        postInvalidate();
    }
}

File: appkotlin/src/main/kotlin/indi/yume/view/appkotlin/view/PwdCheckView.kt
package indi.yume.view.appkotlin.view

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.support.annotation.IntRange
import android.support.annotation.Size
import android.support.annotation.StyleableRes
import android.util.AttributeSet
import android.view.View

import indi.yume.view.appkotlin.R
import indi.yume.view.appkotlin.util.CheckPwdUtil

/**
 * Created by bush2 on 2016/4/5.
 */
class PwdCheckView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {

    @Size(5)
    private val rectPaintList = arrayOf(Paint(), Paint(), Paint(), Paint(), Paint())
    private val backRectPaint = Paint()

    @Size(5)
    private val rectList = arrayOf(Rect(), Rect(), Rect(), Rect(), Rect())
    private var level = 0

    init {

        if (attrs != null)
            init(context.obtainStyledAttributes(attrs, R.styleable.PwdCheckView))

        init()
    }

    private fun init(tArray: TypedArray) {
        val count = tArray.indexCount
        for (i in 0..count - 1)
            setAttr(tArray.getIndex(i), tArray)
        tArray.recycle()
    }

    private fun setAttr(@StyleableRes attrIndex: Int, tArray: TypedArray) {
        when (attrIndex) {
            R.styleable.PwdCheckView_pcv_password -> {
                val pwd = tArray.getString(attrIndex)
                checkLevel(pwd)
            }
        }
    }

    private fun init() {
        for (i in 0..4) {
            rectPaintList[i].isAntiAlias = true
            rectPaintList[i].color = RECT_COLOR_LIST[i]
        }

        backRectPaint.isAntiAlias = true
        backRectPaint.color = BACK_RECT_COLOR
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)

        val cellWithSpace = w / MAX_LEVEL
        val space = cellWithSpace / 6

        val cellWidth = (w - (MAX_LEVEL - 1) * space) / MAX_LEVEL

        var xOffset = 0
        for (i in 0..MAX_LEVEL - 1) {
            val rect = rectList[i]
            rect.set(xOffset, 0, xOffset + cellWidth, h)
            xOffset += cellWidth + space
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        var paint: Paint
        for (i in 0..MAX_LEVEL - 1) {
            if (i > level)
                paint = backRectPaint
            else
                paint = rectPaintList[i]

            canvas.drawRect(rectList[i], paint)
        }
    }

    fun setLevel(@IntRange(from = 0, to = MAX_LEVEL.toLong()) level: Int) {
        this.level = level
        postInvalidate()
    }

    fun checkLevel(pwd: String) {
        this.level = CheckPwdUtil.checkPwdStrong(pwd, 8, 8)
        postInvalidate()
    }

    companion object {
        private const val MAX_LEVEL = 5
        private val RECT_COLOR_LIST = intArrayOf(0xffe11c1b.toInt(), 0xffe0791b.toInt(), 0xffe1ce1b.toInt(), 0xff83e01b.toInt(), 0xff1ce11a.toInt())
        private const val BACK_RECT_COLOR = 0xffeeeeee.toInt()
    }
}
